huffman/src/table.hpp

```diff
@@ -67,18 +67,20 @@ class table
 
   constexpr auto encode_symbols() -> void
   {
+    auto reversed = std::views::reverse(table_);
+
     // precondition, audit
-    assert(std::ranges::is_sorted(table_));
+    assert(std::ranges::is_sorted(reversed));
 
-    const auto total_size = table_.size();
-    auto first = table_.begin();
-    const auto last = table_.end();
+    const auto total_size = reversed.size();
+    auto first = reversed.begin();
+    const auto last = reversed.end();
 
     while (first->node_size() != total_size) {
-      join(first[0], first[to_index(first->node_size())]);
+      join_reversed(first[0], first[to_index(first->node_size())]);
 
-      const auto has_higher_freq = [f = first->frequency()](const auto& n) {
-        return n.frequency() > f;
+      const auto has_higher_freq = [&first](const auto& n) {
```

Comment:
Is this change just for readability?

I guess you're hoping the compiler doesn't call first->frequency() each time and instead in-lines the constant?

Reply:
Before, we need to create a new lambda expression in each iteration of the while loop, as the frequency of the first node changes.

After, the lambda expression doesn't actually need to change since the address of `first` remains constant. We could move this out of the while loop but I assume the compiler could to that fairly easily.

I find it a bit easier to read as line 83 has `frequency()` on both sides of the comparison operator.